### DISCUSSION TITLE
DeepForest ONNX model

### DISCUSSION THREAD

**Discussion by matismatis** [PUTvision/qgis-plugin-deepness#124] [Q&A]
Hello everyone, 

I am Matis, a forestry student interested in GIS/remote sensing. I made my own orthophoto map with a drone, and I want to detect the trees. I wanted to use DeepForest within Deepness for this. The plugin Deepness seemed a perfect plugin for me, as I don't have a lot of python experience.
I tried to convert the model from DeepForest (PyTorch) to ONNX. Here is the link to the model that I want to convert: https://github.com/weecology/DeepForest/blob/main/deepforest/model.py 

I have the code in the attached as a text file. 
[DeepForest_PyTorch_to_ONNX.txt](https://github.com/PUTvision/qgis-plugin-deepness/files/13531040/DeepForest_PyTorch_to_ONNX.txt)

However, I got this error notification from Deepness/QGIS;

''Error! Failed to load the model!
Model may not be usable.
Exception: Model with multiple output layer is not supported! Use only one output layer.''

As I said I am still inexperienced to programming/GIS in general, so sorry if this is a low effort/simple question. Any help would greatly be appreciated!

Edit: I tried adding a ZIP file of the ONNX that I created, but it seems to be too large. 

Best regards,
Matis

**Comment by bartoszptak** [comment]
Hi @matismatis!
It looks like DeepForest uses RetinaNet and FasterRCNN architectures which are not currently supported in the Deepness plugin. They differ from the supported YOLO family in the way they process results from the network. We don't exclude adding support for them in the future, but we don't currently have plans to do it.

**Reply by matismatis** [reply]
Thank you very much! I see, I didn't read the model requirements properly. Would it be possible to use the large dataset that was also used in DeepForest to train a YOLO model? That would save me a lot of work marking trees. 
I tried to acces the guide on making a model in python (./tutorials/detection/cars_yolov7/car_detection__prepare_and_train.ipynb), however, I could not open the file (An error occurred Using nbformat v5.8.0 and nbconvert v7.3.1).

Edit: This is the dataset https://zenodo.org/records/3765872

Again, thank you for your help.

**Comment by fredrikbacke92** [comment]
I did a write up a while ago that might suit you. I also made a model for
detecting trees. You can upload your dataset to roboflow and annotate
there. Then export to ultralytics to train a YOLO model and export to ONNX.
Don't know if your annotations will follow along your dataset but roboflow
has a "magic select" which is powered by Metas segment anything and would
make it much quicker to redo at least. See link below.

qgis-plugin-deepness/tutorials/detection/Deploy_object_detection_model_with_Deepness_without_a_single_line_of_code.pdf
at devel · PUTvision/qgis-plugin-deepness · GitHub
<https://github.com/PUTvision/qgis-plugin-deepness/blob/devel/tutorials/detection/Deploy_object_detection_model_with_Deepness_without_a_single_line_of_code.pdf>

Cheers

Den lör 2 dec. 2023 kl 00:52 skrev matismatis ***@***.***>:

**Reply by aloboa** [reply]
Nice! Do you know if it possible (and if yes, how) to use the mask generated by deepness/Training data export to train a model in YOLOv8?
